internal/transport/grpchttp2/http2bridge.go

```diff
@@ -0,0 +1,214 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpchttp2
+
+import (
+	"io"
+
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/hpack"
+	"google.golang.org/grpc"
+)
+
+const (
+	initHeaderTableSize = 4096
+	maxFrameLen         = 16384
+)
+
+type HTTP2FramerBridge struct {
+	framer *http2.Framer
+	buf    [maxFrameLen]byte
+	pool   grpc.SharedBufferPool
+}
+
+func NewHTTP2FramerBridge(w io.Writer, r io.Reader, maxHeaderListSize uint32) *HTTP2FramerBridge {
+	fr := &HTTP2FramerBridge{
+		framer: http2.NewFramer(w, r),
+		pool:   grpc.NewSharedBufferPool(),
+	}
+
+	fr.framer.SetReuseFrames()
+	fr.framer.MaxHeaderListSize = maxHeaderListSize
+	fr.framer.ReadMetaHeaders = (hpack.NewDecoder(initHeaderTableSize, nil))
+
+	return fr
+}
+
+func (fr *HTTP2FramerBridge) ReadFrame() (Frame, error) {
+	f, err := fr.framer.ReadFrame()
+	if err != nil {
+		return nil, err
+	}
+
+	h := f.Header()
+	hdr := &FrameHeader{
+		Size:     h.Length,
+		Type:     FrameType(h.Type),
+		Flags:    Flag(h.Flags),
+		StreamID: h.StreamID,
+	}
+
+	switch f := f.(type) {
+	case *http2.DataFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.Data())
+		df := &DataFrame{
+			hdr:  hdr,
+			Data: buf,
+		}
+		df.free = func() {
+			fr.pool.Put(&buf)
+			df.Data = nil
```

Comment:
Should setting this to `nil` be done in the free function instead?

Reply:
This is now done by the `mem` package.